ibridges/irodsconnector/sync.py

```diff
@@ -0,0 +1,376 @@
+
+        for subcoll in coll.subcollections:
+            subobjects, subcollections=_get_irods_tree(
+                coll=subcoll,
+                root=root,
+                level=level+1,
+                max_level=max_level,
+                ignore_checksum=ignore_checksum
+                )
+            objects.extend(subobjects)
+            collections.extend(subcollections)
+    else:
+        collections=[]
+
+    return objects, collections
+
+def _create_irods_collections(session: Session,
+                              target: IrodsPath,
+                              collections: list[FolderObject],
+                              dry_run: bool,
+                              copy_empty_folders: bool):
+    if dry_run:
+        print("Will create collection(s):")
+
+    for collection in collections:
+        if collection.is_empty() and not copy_empty_folders:
+            continue
+
+        full_path=target / collection.path
+
+        if dry_run:
+            print(f"  {full_path} {collection.n_files}/{collection.n_folders}")
+        else:
+            _=create_collection(session, str(full_path))
+
+def _create_local_folders(target: Path,
+                          folders: list[FolderObject],
+                          dry_run: bool,
+                          copy_empty_folders: bool):
+    if dry_run:
+        print("Will create folder(s):")
+
+    for folder in folders:
+        if folder.is_empty() and not copy_empty_folders:
+            continue
+
+        full_path=target / Path(folder.path)
+
+        if dry_run:
+            print(f"  {full_path}")
+        else:
+            full_path.mkdir(parents=True, exist_ok=True)
+
+def _copy_local_to_irods(session: Session,   #pylint: disable=too-many-arguments
+                         source: Path,
+                         target: IrodsPath,
+                         files: list[FileObject],
+                         dry_run: bool,
+                         verify_checksum: bool) -> None:
+    if dry_run:
+        print(f"Will upload from '{source}' to '{target}':")
+    else:
+        pbar=tqdm(desc='Uploading', total=sum(x.size for x in files))
+
+    for file in files:
+        source_path=Path(source) / file.path
+        target_path=str(target / file.path)
+        if dry_run:
+            print(f"  {file.path}  {file.size}")
+        else:
+            try:
+                upload(session=session,
+                       local_path=source_path,
+                       irods_path=target_path,
+                       overwrite=True)
+                if verify_checksum:
+                    obj=get_dataobject(session, target_path)
+                    if file.checksum != \
+                            (obj.checksum if len(obj.checksum)>0 else obj.chksum()):
+                        # log.warning("Checksum mismatch after upload: '%s'", target_path)
+                        print("WARNING: Checksum mismatch after upload: '%s'", target_path)
```

Comment:
I personally feel like a printed warning is too little for something as serious as a checksum fail. We already have a lot of arguments, but I think typically it is done by something like on_fail = "skip"/"error"/"warning".

Reply:
Ok, good idea, one extra param won't hurt. Would 'skip' mean ignoring the error or deleting the faulty file?

Reply:
Perhaps for now, we can just support "error" and "warning", with default to "error". The checksum can always be not checked. Another option might be "delete" I suppose.